New: capitalized-comments rule (fixes #6055)

platinumazure opened the pull request:
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[x] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

See issue #6055.

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (http://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

**What changes did you make? (Give an overview)**

Added new rule "capitalized-comments".

**Is there anything you'd like reviewers to focus on?**

I've tried to implement the tests cases from [the JSCS tests](https://github.com/jscs-dev/node-jscs/blob/master/test/specs/rules/require-capitalized-comments.js). It'd be great to confirm I haven't missed any :smile:

I intentionally did not implement "markers" because (1) it's still not clear exactly how they would work in this rule (though it is clear to me that they would need to behave differently than `spaced-comment`); and (2) it's not necessary for JSCS compatibility.

I had to make some decisions regarding a few JSCS rule features and whether they should be default behavior or behind an option.
- Comments that start with URLs are not reported: I made this default behavior.
- Consecutive comments ("textblocks" in JSCS-speak) are not warned if the first comment in the group is valid: I've implemented that as a `tolerateConsecutiveComments` option, off by default.
- Ignoring inline comments: This option was called `inlined` in JSCS, but I'm calling it `ignoreInlineComments` here for clarity.

platinumazure commented:
Labeling "accepted"/"feature"/"rule" to match the issue.

not-an-aardvark commented on docs/rules/capitalized-comments.md:
````diff
@@ -0,0 +1,241 @@
+Examples of **incorrect** code with different line and block comment configuration:
+
+```js
+/* eslint capitalized-comments: ["error", { "block": { "capitalize": "always" }, "line": { "capitalize": "never" } }]
````
You're missing a closing `*/` here.

nzakas replied:
Also missing closing `*/` here.

not-an-aardvark commented on lib/rules/capitalized-comments.js:
```diff
@@ -0,0 +1,296 @@
+
+const ALWAYS_MESSAGE = "Comments should not begin with a lowercase character",
+    NEVER_MESSAGE = "Comments should not begin with an uppercase character",
+    DEFAULT_IGNORE_PATTERN = /^\s*(?:eslint|istanbul|jscs|jshint|globals?|exported)\b/,
```
Do we also want to exclude comments like `TODO` by default? I think these are almost always capitalized.

not-an-aardvark commented on tests/lib/rules/capitalized-comments.js:
```diff
@@ -0,0 +1,1012 @@
+            }]
+        },
+
+        // ignoreInlineComments should only allow inline comments to pass
```
Can you add a test to ensure that inline comments are *not* ignored if the option is omitted/set to false?

nzakas commented on docs/rules/capitalized-comments.md:
```diff
@@ -0,0 +1,241 @@
+* `ignorePattern`: A string representing a regular expression pattern of words that should be ignored by this rule. If the first word of a comment matches the pattern, this rule will not report that comment.
+    * Note that the following words are always ignored by this rule: `["jscs", "jshint", "eslint", "istanbul", "global", "globals", "exported"]`.
+* `ignoreInlineComments`: If this is `true`, the rule will not report on comments in the middle of code. By default, this is `false`.
+* `tolerateConsecutiveComments`: If this is `true`, the rule will not report on a comment which violates the rule, as long as the comment immediately follows a comment which is also not reported.
```
Agreed. We should try to keep option names with the same verb.

platinumazure replied:
For sure! Not sure what I was thinking earlier but it's obvious now. I'll go with `ignore` due to fewer characters.

nzakas commented on docs/rules/capitalized-comments.md:
```diff
@@ -0,0 +1,241 @@
+# enforce or disallow capitalization of the first letter of a comment (capitalized-comments)
+
+This rule was ported from JSCS' [requireCapitalizedComments](http://jscs.info/rule/requireCapitalizedComments) and [disallowCapitalizedComments](http://jscs.info/rule/disallowCapitalizedComments) rules.
```
We should still have a standalone description here.

nzakas commented on docs/rules/capitalized-comments.md:
```diff
@@ -0,0 +1,241 @@
+
+### Options
+
+This rule has one option: either a string value `"always"` or `"never"` for simple configuration, or an object containing all possible configuration parameters for the rule.
```
I agree. 👍

nzakas commented on lib/util/patterns/L.js:
```diff
@@ -0,0 +1,10 @@
+/**
```
Did you grab this from JSCS? If so, we need to be sure to include proper licensing information here.

Also, can you rename this file? It doesn't follow our conventions of all lowercase and dash-cased. Maybe `letters.js`?

platinumazure replied:
I can rename if you like, sure. The name was chosen (in JSCS) to match the shorthand POSIX character class `\p{L}`.

Can you give me some guidance on noting the license correctly? Thanks!

platinumazure commented:
Regarding the schema:

The one challenge is how to handle the line/block configuration. If we just remove the `capitalize` option from the object and make it a string, then there's no way to require capitalization for line comments and require lowercase for block comments (or vice versa). In that sense, I went for a simple implementation which says you can use the string object for super easy configuration, otherwise you have to use the object.

So there are a couple of options, I suppose:

* Require the string option in all cases except severity-only configuration; don't allow the `capitalize` key in the simple object configuration case, but allow the key in the line/block configuration case.

    ```js
    {
        "capitalized-comments": ["error"],
        "capitalized-comments": ["error", "always"],
        "capitalized-comments": ["error", "always", { "ignoreInlineComments": true }],  // capitalize key forbidden here
        "capitalized-comments": ["error", "always", { "line": { "capitalize": "never", "ignoreInlineComments": true } }]
    }
    ```

* Require the string option except in line/block configuration-- in that configuration, the schema must just be the object.

    ```js
    {
        "capitalized-comments": ["error"],
        "capitalized-comments": ["error", "always"],
        "capitalized-comments": ["error", "always", { "ignoreInlineComments": true }],  // capitalize key forbidden here
        "capitalized-comments": ["error", { "line": { "capitalize": "never", "ignoreInlineComments": true } }]  // Here, block comments get capitalize: "always" automatically
    }
    ```

platinumazure commented:
What I've just pushed:

* All the minor documentation issues
* Added a description (very brief, let me know if it should be longer)
* Changed `tolerateConsecutiveComments` to `ignoreConsecutiveComments`
* Added requested `ignoreInlineComments` negative tests

Left to do:

* Agree on (and then implement) a schema
* Rename and add licensing information for the Unicode letter pattern file

not-an-aardvark commented:
Hmm, you have a good point about separate line/block configurations. I guess it just seems redundant to have two separate schemas for accomplishing the same thing (assuming the user doesn't need any of the other options).

Did JSCS allow you to configure line and block comments separately (e.g. requiring uppercase for block comments, and lowercase for line comments)? I'm thinking it might be simpler if the uppercase/lowercase preference applied to both line and block comments, since I'm not sure why a project would want to require capitalization for one type but not the other. That would allow us to always stick with a string option, and use an object for additional configuration.

nzakas commented:
Just a heads up: we have moved to a new CLA checker on pull requests. Even if you've previously signed our CLA, we will need to you sign the new one. To do so, look at the status checks for **licence/cla** and click the "Details" link. Sorry for the inconvenience.

platinumazure commented:
@not-an-aardvark 

> Did JSCS allow you to configure line and block comments separately (e.g. requiring uppercase for block comments, and lowercase for line comments)? I'm thinking it might be simpler if the uppercase/lowercase preference applied to both line and block comments, since I'm not sure why a project would want to require capitalization for one type but not the other. That would allow us to always stick with a string option, and use an object for additional configuration.

JSCS did not allow this, so I am inclined to agree with you that we don't need to allow it either. Given that decision, it makes sense to extract the always/never option out of the object. I'll implement that in the next few days and then ping you again.

@nzakas I'm still looking for guidance on how to determine and represent a license for code that was in JSCS, when you have a chance.

platinumazure commented:
@not-an-aardvark @nzakas Schema has been modified per your recommendations. I've also renamed the file that needed to be renamed. Except for the license needed in the pattern file, this should be good for another round of review. Thanks!

not-an-aardvark commented on docs/rules/capitalized-comments.md:
````diff
@@ -0,0 +1,243 @@
+Examples of **correct** code with the `"ignorePattern"` option set to `"pragma":
+
+```js
+/* eslint capitalized-comments: ["error", { "capitalize": "always", "ignorePattern": "pragma" }]
````
It looks like this inline comment is still using the old schema. (It's also missing a `*/`.)

not-an-aardvark commented on docs/rules/capitalized-comments.md:
````diff
@@ -0,0 +1,243 @@
+Examples of **correct** code with the `"ignoreInlineComments"` option set to `true`:
+
+```js
+/* eslint capitalized-comments: ["error", { "capitalize": "always", "ignoreInlineComments": true }] */
````
Same with this one

not-an-aardvark commented on docs/rules/capitalized-comments.md:
````diff
@@ -0,0 +1,243 @@
+Examples of **correct** code with `ignoreConsecutiveComments` set to `true`:
+
+```js
+/* eslint capitalize-comments: ["error", { "capitalize": "always", "ignoreConsecutiveComments": true }] */
````
and this one

not-an-aardvark commented on docs/rules/capitalized-comments.md:
````diff
@@ -0,0 +1,243 @@
+Examples of **incorrect** code with different line and block comment configuration:
+
+```js
+/* eslint capitalized-comments: ["error", { "block": { "capitalize": "always" }, "line": { "capitalize": "never" } }] */
````
and this one

not-an-aardvark commented on docs/rules/capitalized-comments.md:
```diff
@@ -0,0 +1,243 @@
+    "capitalized-comments": [
+        "error",
+        {
+            "line": {
```
This should also use the updated schema.

platinumazure commented:
Doh, you're right, I completely forgot the docs. Sorry.

On Nov 6, 2016 9:15 PM, "Teddy Katz" <notifications@github.com> wrote:

> *@not-an-aardvark* requested changes on this pull request.
>
> The code LGTM, but I think the docs need to be updated for the new schema.
> ------------------------------
>
> In docs/rules/capitalized-comments.md
> <https://github.com/eslint/eslint/pull/7415#pullrequestreview-7349285>:
>
> > +// lowercase comment
> +
> +// 1. Non-letter at beginning of comment
> +
> +// 丈 Non-Latin character at beginning of comment
> +
> +```
> +
> +#### `ignorePattern`
> +
> +The `ignorePattern` object takes a string value, which is used as a regular expression applied to the first word of a comment.
> +
> +Examples of **correct** code with the `"ignorePattern"` option set to `"pragma":
> +
> +```js
> +/* eslint capitalized-comments: ["error", { "capitalize": "always", "ignorePattern": "pragma" }]
>
> It looks like this inline comment is still using the old schema. (It's
> also missing a */.)
> ------------------------------
>
> In docs/rules/capitalized-comments.md
> <https://github.com/eslint/eslint/pull/7415#pullrequestreview-7349285>:
>
> > +/* eslint capitalized-comments: ["error", { "capitalize": "always", "ignorePattern": "pragma" }]
> +
> +function foo() {
> +    /* pragma wrap(true) */
> +}
> +
> +```
> +
> +#### `ignoreInlineComments`
> +
> +Setting the `ignoreInlineComments` option to `true` means that comments in the middle of code (with a token on the same line as the beginning of the comment, and another token on the same line as the end of the comment) will not be reported by this rule.
> +
> +Examples of **correct** code with the `"ignoreInlineComments"` option set to `true`:
> +
> +```js
> +/* eslint capitalized-comments: ["error", { "capitalize": "always", "ignoreInlineComments": true }] */
>
> Same with this one
> ------------------------------
>
> In docs/rules/capitalized-comments.md
> <https://github.com/eslint/eslint/pull/7415#pullrequestreview-7349285>:
>
> > +```js
> +/* eslint capitalized-comments: ["error", { "capitalize": "always", "ignoreInlineComments": true }] */
> +
> +function foo(/* ignored */ a) {
> +}
> +
> +```
> +
> +#### `ignoreConsecutiveComments`
> +
> +If the `ignoreConsecutiveComments` option is set to `true`, then comments which otherwise violate the rule will not be reported as long as they immediately follow a comment which did not violate the rule. This can be applied more than once.
> +
> +Examples of **correct** code with `ignoreConsecutiveComments` set to `true`:
> +
> +```js
> +/* eslint capitalize-comments: ["error", { "capitalize": "always", "ignoreConsecutiveComments": true }] */
>
> and this one
> ------------------------------
>
> In docs/rules/capitalized-comments.md
> <https://github.com/eslint/eslint/pull/7415#pullrequestreview-7349285>:
>
> > +```
> +
> +Examples of **incorrect** code with different line and block comment configuration:
> +
> +```js
> +/* eslint capitalized-comments: ["error", { "block": { "capitalize": "always" }, "line": { "capitalize": "never" } }]
> +
> +// Capitalized line comment, this is incorrect
> +/* lowercased block comment, this is incorrect too */
> +
> +```
> +
> +Examples of **incorrect** code with different line and block comment configuration:
> +
> +```js
> +/* eslint capitalized-comments: ["error", { "block": { "capitalize": "always" }, "line": { "capitalize": "never" } }] */
>
> and this one
> ------------------------------
>
> In docs/rules/capitalized-comments.md
> <https://github.com/eslint/eslint/pull/7415#pullrequestreview-7349285>:
>
> > +/* and this one is valid as well; */
> +/*
> + * in fact, even if any of these are multi-line, that is fine too.
> + */
> +```
> +
> +### Using Different Options for Line and Block Comments
> +
> +If you wish to have a different configuration for line comments and block comments, you can do so by using two different object configurations:
> +
> +```json
> +{
> +    "capitalized-comments": [
> +        "error",
> +        {
> +            "line": {
>
> This should also use the updated schema.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/eslint/pull/7415#pullrequestreview-7349285>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AARWepk8oGJQDohJxdWfV_R0ApEh9qHlks5q7omogaJpZM4KcdaG>
> .
>

platinumazure commented:
@not-an-aardvark Wrote some quick docs changes while waiting to board my aircraft, please do let me know if I missed anything.

not-an-aardvark commented on docs/rules/capitalized-comments.md:
````diff
@@ -0,0 +1,240 @@
+Examples of **correct** code with the `"ignorePattern"` option set to `"pragma":
+
+```js
+/* eslint capitalized-comments: ["error", "always", { "ignorePattern": "pragma" }]
````
Missing `*/` here

not-an-aardvark replied:
Oh also, I missed it before submitting this review but there's also a missing `` ` `` after "pragma" right before this code block.

not-an-aardvark commented on docs/rules/capitalized-comments.md:
````diff
@@ -0,0 +1,240 @@
+Examples of **correct** code with different line and block comment configuration:
+
+```js
+/* eslint capitalized-comments: ["error", "always", { "block": { "ignorePattern": "blockignore" } }]
````
Missing `*/` here

platinumazure commented:
@not-an-aardvark Thanks for reviewing, sorry for missing those. I'll have to fix those when I get home in a few hours.

vitorbal commented:
This one looks almost ready to go. Great job, @platinumazure!

Can anyone provide some guidance on what is needed to [include proper licensing information](https://github.com/eslint/eslint/pull/7415#discussion_r84557217) for the JSCS chunk of the code? Does it mean adding original `@author` information and a link to the JSCS license?

platinumazure commented:
By the way, the file in question was generated- if I regenerate it and
submit it under my own name, is that preferable?

On Nov 11, 2016 12:42 PM, "Vitor Balocco" <notifications@github.com> wrote:

> This one looks almost ready to go. Great job, @platinumazure
> <https://github.com/platinumazure>!
>
> Can anyone provide some guidance on what is needed to include proper
> licensing information
> <https://github.com/eslint/eslint/pull/7415#discussion_r84557217> for the
> JSCS chunk of the code? Does it mean adding original @author information
> and a link to the JSCS license?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/eslint/pull/7415#issuecomment-260025580>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AARWempEUrsuzT1CY6JggYGsD6XHkUiaks5q9LcEgaJpZM4KcdaG>
> .
>

platinumazure commented:
@not-an-aardvark I believe I've addressed your latest comments; please let me know if I have missed something.

@eslint/eslint-tsc I just need some guidance on how to add licensing information to `lib/util/patterns/letters.js` since it came from JSCS. Then hopefully this will be ready to go!

nzakas commented on docs/rules/capitalized-comments.md:
```diff
@@ -0,0 +1,240 @@
+
+The purpose of this rule is to enforce consistent casing of the first word of every comment in your codebase.
+
+This rule was ported from JSCS' [requireCapitalizedComments](http://jscs.info/rule/requireCapitalizedComments) and [disallowCapitalizedComments](http://jscs.info/rule/disallowCapitalizedComments) rules.
```
This sentence isn't necessary. Just use the "Compatibility" section later on.

nzakas commented on docs/rules/capitalized-comments.md:
```diff
@@ -0,0 +1,240 @@
+# enforce or disallow capitalization of the first letter of a comment (capitalized-comments)
+
+The purpose of this rule is to enforce consistent casing of the first word of every comment in your codebase.
```
This section should be about the concept that the rule addresses, not about the rule itself. See, for example, http://eslint.org/docs/rules/brace-style

nzakas commented on lib/util/patterns/letters.js:
```diff
@@ -0,0 +1,10 @@
+ * @fileoverview Pattern for detecting any letter (even letters outside of
+ * ASCII).
+ * @author Kevin Partington
+ */
```
Since this file was generated using a tool, I'd just throw a sentence in here saying that this file was generated by whatever tool generated it and not to edit by hand.

platinumazure replied:
Thanks, should I change the author information to match who generated it for JSCS? (And what about licensing?)

Or do you want me to regenerate this myself?

Sorry, the author field is misleading right now.

platinumazure commented:
@nzakas Hopefully ready for review again. I decided to add the JSCS license to lib/util/patterns/letters.js, let me know if you want it gone.

nzakas commented:
I don't think it's necessary, but it's fine to leave it in.
